sdk/core/azure-core/src/main/java/com/azure/core/util/logging/LoggingEventBuilder.java

```diff
@@ -0,0 +1,289 @@
+    }
+
+    /**
+     * Adds key with String value pair to the context of current log being created.
+     *
+     * <p><strong>Code samples</strong></p>
+     *
+     * <p>Adding string value to logging event context.</p>
+     *
+     * <!-- src_embed com.azure.core.util.logging.clientlogger.atInfo -->
+     * <pre>
+     * logger.atInfo&#40;&#41;
+     *     .addKeyValue&#40;&quot;key&quot;, &quot;value&quot;&#41;
+     *     .log&#40;&quot;A formattable message. Hello, &#123;&#125;&quot;, name&#41;;
+     * </pre>
+     * <!-- end com.azure.core.util.logging.clientlogger.atInfo -->
+     *
+     * @param key String key.
+     * @param value String value.
+     * @return The updated {@code LoggingEventBuilder} object.
+     */
+    public LoggingEventBuilder addKeyValue(String key, String value) {
+        if (this.isEnabled) {
+            addKeyValueInternal(key, value);
+        }
+
+        return this;
+    }
+
+    /**
+     * Adds key with boolean value to the context of current log being created.
+     *
+     * @param key String key.
+     * @param value boolean value.
+     * @return The updated {@code LoggingEventBuilder} object.
+     */
+    public LoggingEventBuilder addKeyValue(String key, boolean value) {
+        if (this.isEnabled) {
+            addKeyValueInternal(key, value);
+        }
+        return this;
+    }
+
+    /**
+     * Adds key with long value to the context of current log event being created.
+     *
+     * <p><strong>Code samples</strong></p>
+     *
+     * <p>Adding an integer value to logging event context.</p>
+     *
+     * <!-- src_embed com.azure.core.util.logging.clientlogger.atverbose.addKeyValue#primitive -->
+     * <pre>
+     * logger.atVerbose&#40;&#41;
+     *     .addKeyValue&#40;&quot;key&quot;, 1L&#41;
+     *     .log&#40;&#40;&#41; -&gt; String.format&#40;&quot;Param 1: %s, Param 2: %s, Param 3: %s&quot;, &quot;param1&quot;, &quot;param2&quot;, &quot;param3&quot;&#41;&#41;;
+     * </pre>
+     * <!-- end com.azure.core.util.logging.clientlogger.atverbose.addKeyValue#primitive -->
+     *
+     * @param key String key.
+     * @param value long value.
+     * @return The updated {@code LoggingEventBuilder} object.
+     */
+    public LoggingEventBuilder addKeyValue(String key, long value) {
+        if (this.isEnabled) {
+            addKeyValueInternal(key, value);
+        }
+        return this;
+    }
+
+    /**
+     * Adds key with String value supplier to the context of current log event being created.
+     *
+     * @param key String key.
+     * @param valueSupplier String value supplier function.
+     * @return The updated {@code LoggingEventBuilder} object.
+     */
+    public LoggingEventBuilder addKeyValue(String key, Supplier<String> valueSupplier) {
+        if (this.isEnabled) {
+            if (this.context == null) {
+                this.context = new ArrayList<>();
+            }
+
+            this.context.add(new ContextKeyValuePair(key, valueSupplier));
```

Comment:
I wonder if we could save a little bit of GC here by replacing the `ArrayList` and the `ContextKeyValuePair` by instead using the `Context` API that already exists in azure-core?

Reply:
Do you mean that we can potentially pass Context object and iterate over keys here?

E.g. in `amqp-core`, it looks like we don't store things like `connectionId` or `linkName` in the context. In EventHubs SDK we don't store those in `Context`, we rather use `PartitionContext`, e.g. [here](https://github.com/Azure/azure-sdk-for-java/blob/main/sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/PartitionPumpManager.java#L218).
I've been looking for a bit, but could not find an example when we pass logging-related context through the `Context`. Also, assuming we start doing it, we pass all kinds of things (and customers too) through it, we'd have to come up with some conventions if we want to take Context as a parameter here and get stuff from it.

Or did I misunderstand you?

E.g. here https://github.com/Azure/azure-sdk-for-java/blob/main/sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/ReactorReceiver.java#L86

Reply:
Not quite - what I was meaning was that instead of creating a List / ContextKeyValuePair internally, we create an internal Context and populate that with the key / value pairs instead. So, we would use the Context type, but it would not be in any way related to the existing usages of Context.

Reply:
Makes sense, I still don't understand the part about GC - with our `Context` we still allocate a new `Context` object on each add. also the only way to iterate over our context is to create a map and copy data from the context to this map. So it seems we'd only be saving on possible array resize?

I can get rid of the `KeyValuePair` and store keys interleaved with values in a single array list, it will remove extra allocation.

Reply:
My mistake - I thought `Context` had a 'getParent' or similar API to move through the linked list, but that doesn't exist. You can disregard my comment as it yields no win.